lib/Optimizer/IROptimizer.cpp

```diff
@@ -1123,6 +1142,106 @@ static void eliminateDeadStores(IRFunction &M) {
   eraseInstructions(M, erasedInstructions);
 }
 
+/// \returns true if the first dimension in \p offsets has a non-zero value.
+bool isOnlyOffsetInFirstDim(llvm::ArrayRef<size_t> offsets) {
+  if (offsets.size() > 1) {
+    for (size_t i = 1; i < offsets.size(); ++i) {
+      if (offsets[i] != 0) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+/// \returns true if the dimensions of \p sourceDims, other than the first
+/// dimension, all equal the dimensions in \p destDims.
+bool allButFirstDimsEqual(llvm::ArrayRef<size_t> sourceDims,
+                          llvm::ArrayRef<size_t> destDims) {
+  assert(sourceDims.size() == destDims.size() &&
+         "Source and dest dims must have same number of dims.");
+  if (sourceDims.size() > 1) {
+    for (size_t i = 1; i < sourceDims.size(); ++i) {
+      if (sourceDims[i] != destDims[i]) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+/// Replace InsertTensors that are only offset in the first dimension with
+/// writing directly into the destination using TensorViews with the same
+/// offsets. This is possible because this means the underlying memory is
+/// contiguous in this case.
+void optimizeInserts(IRFunction &M) {
+  auto &instrs = M.getInstrs();
+  Instructions erasedInstructions;
+  IRBuilder B(&M);
+  for (auto it = instrs.begin(), e = instrs.end(); it != e; ++it) {
+    auto *I = *it;
+    // Look for compatible InsertTensors.
+    if (auto *ITI = dyn_cast<InsertTensorInst>(I)) {
+      // TVI with offsets currently only works for this optimization if the
+      // insert is only into the first dimension. This is because the tensor
+      // memory must be contiguous.
+      if (!isOnlyOffsetInFirstDim(ITI->getOffsets())) {
+        continue;
+      }
+      if (auto *insertSourceAAI =
+              dyn_cast<AllocActivationInst>(ITI->getSrc())) {
+        // Assume 3 users of the source AAI (ITI, the original instruction that
+        // writes into it, and a dealloc). This is the pattern generated by an
+        // IRGen'd ConcatNode.
+        if (insertSourceAAI->getNumUsers() == 3) {
```

Comment:
It is probably not enough to check only the number of uses. You may need to check exactly what the comment above states, i.e. that there is an alloc, the write and the dealloc.

Reply:
Above I already have checked if the insert's source is an `AllocActivationInst` which accounts for one use, and then assuming well formedness there must be a second use that is `DeallocActivationInst`. So this means the third use must be the insert here. Do you think explicitly verifying `DeallocActivationInst` is necessary?